CMakeLists.txt

```diff
@@ -160,18 +168,46 @@ SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
 SET(CMAKE_INSTALL_RPATH "$ORIGIN:$ORIGIN/lib:$ORIGIN/lib/plugins:$ORIGIN/../lib:$ORIGIN/../lib/plugins")
 
 SET(BLA_STATIC ON)
+if(NOT DEFINED ZLIB_USE_STATIC_LIBS)
+  SET(ZLIB_USE_STATIC_LIBS TRUE)
+endif()
 if(NOT DEFINED OPENSSL_USE_STATIC_LIBS)
   SET(OPENSSL_USE_STATIC_LIBS TRUE)
 endif()
 
-if (CUDAQ_BUILD_SELFCONTAINED)
-  # Using -static-libgcc -static-libstdc++ offers finer control here if needed.
+if(NOT DEFINED CUDAQ_ENABLE_STATIC_LINKING)
+  SET(CUDAQ_ENABLE_STATIC_LINKING $ENV{CUDAQ_ENABLE_STATIC_LINKING})
+endif()
+if (CUDAQ_ENABLE_STATIC_LINKING)
+  # Using -static-libgcc -static-libstdc++ offers more fine grained control here if needed.
   # At the time of writing, there wasn't any benefit to using those over -static.
+  # However, while this so far seems to be working reasonably well, this will
+  # unfortunately also link glibc statically, which in general is likely to cause
+  # issues sooner or later. A better option might be to include the required GNU C Library
+  # and have the nvq++ compiler dynamically pick the newer one of the included or system one.
+  # See also these posts:
+  # - https://blog.gibson.sh/2017/11/26/creating-portable-linux-binaries/
+  # - https://www.jertype.com/upgrading-glibc/
+  # If the final application loads multiple dynamic libraries, and two or more
+  # of those dynamic libraries are linked to different versions of the same static
+  # libraries, then the different copies of those static libraries will conflict
+  # with each other. If we keep linking everything statically, we should at least set 
+  # CMAKE_<LANG>_VISIBILITY_PRESET to hidden and CMAKE_VISIBILITY_INLINES_HIDDEN to 1, 
+  # to avoid some of these issues.
```

Comment:
The premise of all this is that `nvq++` *is* `clang++` under the hood and needs to be installed right along side of an identical `clang++`. That should be the default and fully supported. (But it is granted, that may also be problematic on some systems.)

Reply:
My concern here are the runtime libraries, which are linked with other user libraries (i.e. libraries that we don't build and have no control over) using nvq++ to build applications. My understanding based on what I read is that if our runtime library statically links glibc without hiding symbols as much as possible it can cause conflicts when it is then linked with a user library that may contain the same symbols (depending on how that library is built). I am not sure how common that would be.